JPEGImageCompression/image_compression.py

```diff
@@ -1,19 +1,145 @@
+        @param: buff - defines how much space there will be between each cube
+        """
+        max_color_res = 256
+        color_resolution = color_res
+        discrete_ratio = max_color_res // color_resolution
+
+        side_length = cube_side_length
+        offset = side_length + buff
+        cubes = []
+
+        for i in range(color_resolution):
+            for j in range(color_resolution):
+                for k in range(color_resolution):
+
+                    color = rgb_to_hex(
+                        (
+                            (i * discrete_ratio) / 255,
+                            (j * discrete_ratio) / 255,
+                            (k * discrete_ratio) / 255,
+                        )
+                    )
+                    cubes.append(
+                        Cube(
+                            side_length=side_length, fill_color=color, fill_opacity=1
+                        ).shift((LEFT * i + UP * j + OUT * k) * offset)
+                    )
+
+        cubes_vg = Group(*cubes)
+        return cubes_vg
+
+    def create_yuv_cube(self, color_res=8, cube_side_length=0.1, buff=0.05):
+        """
+        Creates a YCbCr cube composed of many smaller cubes. The `color_res` argument defines
+        how many cubes there will be to represent the full spectrum, particularly color_res ^ 3.
+        Works exactly like `create_rgb_cube` but for YCrCb colors.
+
+        It is recommended that color_res is a power of two.
+
+        @param: color_res - defines the number of cubes in every dimension. Higher values yield a finer
+        representation of the space, but are very slow to deal with.
+        @param: cube_side_length - defines the side length of each individual cube
+        @param: buff - defines how much space there will be between each cube
+        """
+
+        max_color_res = 256
+        color_resolution = color_res
+        discrete_ratio = max_color_res // color_resolution
+
+        side_length = cube_side_length
+        offset = side_length + buff
+        cubes = []
+
+        for i in range(color_resolution):
+            for j in range(color_resolution):
+                for k in range(color_resolution):
+
+                    r = i * discrete_ratio
+                    g = j * discrete_ratio
+                    b = k * discrete_ratio
+
+                    y, cb, cr = self.rgb2ycbcr(r, g, b)
+
+                    color = rgb_to_hex(
+                        (
+                            (y) / 255,
+                            (cb) / 255,
+                            (cr) / 255,
+                        )
+                    )
+
+                    cubes.append(
+                        Cube(
+                            side_length=side_length, fill_color=color, fill_opacity=1
+                        ).shift((LEFT * i + UP * j + OUT * k) * offset)
+                    )
+
+        cubes_vg = Group(*cubes)
+        return cubes_vg
```

Comment:
One overarching comment here that I think is a good learning experience in general software engineering:

Notice how the functions `create_rgb_cube`and `create_yuv_cube` are actually essentially the same function with only one difference. In the yuv function, you have one additional conversion from rgb to ycbcr. As a more general implementation, it is possible to instead define a function

`def create_color_space_cube(self, format, color_res=8, cube_side_length=0.1, buff=0.05)`

And what's nice about this is you could define format as a string for simplicity and then do conditional logic in the function to get the right hex values. OR what I suggest which is really a good way to do it in my opinion is to actually define a function for getting RGB hex and getting YUV hex and pass that into the function.

So you could call the function like
`self.create_color_space_cube(rgbtohex)` to get RGB cubes (inside the function, you could then call the function argument `format` to get the color) and similarly you can call `self.create_color_space_cube(yuvtohex)` to get YUV cubes where now `format` is bound to separate function that gets you colors for the YUV space.

Let me know if you have any questions. This basically utilizes the fact that in Python, you can pass in functions as arguments to a method/another function.